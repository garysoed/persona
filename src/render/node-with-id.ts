export const __id = Symbol('id');

export type NodeWithId<N extends Node> = N & {[__id]: {}};
