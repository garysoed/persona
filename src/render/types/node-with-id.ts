import {hasPropertiesType, instanceofType, intersectType, notType, undefinedType} from 'gs-types';

export const __id = Symbol('id');

export type NodeWithId<N extends Node> = N & {[__id]: unknown};

export const NODE_WITH_ID_TYPE = intersectType([
  instanceofType(Node),
  hasPropertiesType({
    [__id]: notType(undefinedType),
  }),
]);

export function equalNodes(a: NodeWithId<Node>, b: NodeWithId<Node>): boolean {
  return a[__id] === b[__id];
}