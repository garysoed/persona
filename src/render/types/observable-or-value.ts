import {Observable, of as observableOf} from 'rxjs';

export type ObservableOrValue<T> = T|Observable<T>;

export function normalize<T>(value: ObservableOrValue<T>): Observable<T> {
  return value instanceof Observable ? value : observableOf(value);
}
