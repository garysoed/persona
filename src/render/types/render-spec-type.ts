export enum RenderSpecType {
  CUSTOM_ELEMENT,
  ELEMENT,
  HTML,
  NODE,
  TEXT_NODE,
}