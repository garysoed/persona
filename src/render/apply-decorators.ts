import {EMPTY, merge, Observable, of as observableOf} from 'rxjs';
import {mapTo} from 'rxjs/operators';

import {NodeWithId} from './node-with-id';


export type Decorator<N> = (node: N) => Observable<unknown>;

export function applyDecorators<N extends NodeWithId<Node>>(
    node: N,
    ...decorators: ReadonlyArray<Decorator<N>>
): Observable<N> {
  const obs$list = decorators.map(decorator => {
    return decorator(node).pipe(mapTo(EMPTY));
  });
  return merge(
      observableOf(node),
      ...obs$list,
  );
}