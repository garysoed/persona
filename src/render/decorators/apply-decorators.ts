import {EMPTY, merge, Observable, of as observableOf} from 'rxjs';
import {switchMapTo} from 'rxjs/operators';

import {NodeWithId} from '../node-with-id';


export type Decorator<N extends NodeWithId<Node>> = (node: N) => Observable<unknown>;

export function applyDecorators<N extends NodeWithId<Node>>(
    node: N,
    ...decorators: ReadonlyArray<Decorator<N>>
): Observable<N> {
  const obs$list: ReadonlyArray<Observable<N>> = decorators.map(decorator => {
    return decorator(node).pipe(switchMapTo(EMPTY));
  });
  return merge<N>(observableOf(node), ...obs$list);
}