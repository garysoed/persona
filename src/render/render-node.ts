import {Observable} from 'rxjs';

import {renderBase} from './render-base';
import {RenderNodeSpec} from './types/render-node-spec';


export function renderNode<N extends Node>(spec: RenderNodeSpec<N>): Observable<N> {
  return renderBase(spec, spec.node);
}
