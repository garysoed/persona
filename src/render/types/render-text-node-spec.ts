import {Observable} from 'rxjs';

import {RenderSpecType} from './render-spec-type';


interface Input {
  readonly textContent: Observable<string>;
}

export interface RenderTextNodeSpec extends Input {
  readonly type: RenderSpecType.TEXT_NODE;
  readonly textContent: Observable<string>;
}

export function renderTextNode(input: Input): RenderTextNodeSpec {
  return {
    type: RenderSpecType.TEXT_NODE,
    ...input,
  };
}
