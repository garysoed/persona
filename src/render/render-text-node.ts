import {defer, EMPTY, Observable} from 'rxjs';
import {switchMapTo, tap} from 'rxjs/operators';

import {PersonaContext} from '../core/persona-context';

import {applyDecorators, Decorator} from './apply-decorators';
import {NodeWithId} from './node-with-id';
import {renderNode} from './render-node';
import {normalize} from './types/observable-or-value';
import {RenderSpecType} from './types/render-spec-type';
import {RenderTextNodeSpec} from './types/render-text-node-spec';


export function renderTextNode(
    spec: RenderTextNodeSpec,
    context: PersonaContext,
): Observable<NodeWithId<Text>> {
  return defer(() => {
    const ownerDocument = context.shadowRoot.ownerDocument;
    if (!ownerDocument) {
      throw new Error('No owner documents found');
    }
    const node = ownerDocument.createTextNode('');

    return renderNode({
      ...spec,
      type: RenderSpecType.NODE,
      node,
      decorator: node => {
        const decorators: Array<Decorator<NodeWithId<Text>>> = [
          // TODO: Dedupe this.
          node => normalize(spec.text).pipe(
              tap(text => {
                node.textContent = text;
              }),
              switchMapTo(EMPTY),
          ),
        ];
        if (spec.decorator) {
          decorators.push(spec.decorator);
        }
        return applyDecorators(node, ...decorators);
      },
    });
  });
}
