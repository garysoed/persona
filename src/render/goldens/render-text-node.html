<test-host><gst-fake-shadow><!-- #ref -->textContent</gst-fake-shadow></test-host>